server/src/server.js

```diff
@@ -1070,13 +1074,20 @@ const END_POINT_SEPARATOR = "|";
 
 // Get OAuth client params for the client-side authorization flow.
 app.get("/api/fxa-oauth/login/*", function(req, res, next) {
-  if (!req.deviceId) {
+  /*if (!req.deviceId) {
     next(errors.missingSession());
     return;
-  }
+  }*/
   randomBytes(32).then(stateBytes => {
     const state = stateBytes.toString("hex");
-    return setState(req.deviceId, state).then(inserted => {
+    // FIXME: should be a different name than "auth":
+    const cookies = new Cookies(req, res, {keys: dbschema.getKeygrip("auth")});
+    const stateId = req.deviceId || String(Math.random());
+    if (!req.deviceId) {
+      cookies.set("fxaState", stateId, {signed: true});
+    }
+    console.log("Using oauth state", stateId);
+    return setState(stateId, state).then(inserted => {
```

Comment:
Version 27 db support: Logic here assumes a random string as deviceId. If the authentication fails and the data is not deleted in checkstate inside confirm-login route handler, reverting db to 27 will fail to apply stateid_deviceid fk constraint

Reply:
@ianb thoughts on how to fix this issue of fail to revert to db 27 and earlier because of numeric state value that violate fk constraint

Reply:
#4777 could be one fix but in the fix proposed we are allowing a week window